File: Array/src/main/python/solution560.py
from typing import List


class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:

        preSum = [0] * (len(nums) + 1)

        for i in range(1, len(nums) + 1):
            preSum[i] = preSum[i - 1] + nums[i - 1]

        ans = 0

        # 超时了
        # for i in range(1, len(nums) + 1):
        #     for j in range(0, i):
        #         if preSum[i] - preSum[j] == k:
        #             ans += 1

        # 先存一下，避免超时
        cnt_map = {}

        for i in preSum:
            if cnt_map.get(i - k):
                ans += cnt_map.get(i - k)

            cnt = cnt_map.setdefault(i, 0)
            cnt += 1
            cnt_map[i] = cnt

        return ans

File: Array/src/main/java/prefix/Solution560.java
package prefix;

import java.util.HashMap;

public class Solution560 {
    public int subarraySum(int[] nums, int k) {
        int[] preSums = new int[nums.length + 1];

        for (int i = 1; i < nums.length + 1; i++) {
            preSums[i] = preSums[i - 1] + nums[i - 1];
        }

        HashMap<Integer, Integer> cntMap = new HashMap<>();
        int ans = 0;

        for (int preSum : preSums) {
            if (cntMap.containsKey(preSum - k)) {
                ans = ans + cntMap.get(preSum - k);
            }

            int cnt = cntMap.getOrDefault(preSum, 0);
            ++cnt;
            cntMap.put(preSum, cnt);
        }

        return ans;

    }

    public static void main(String[] args) {
        Solution560 s = new Solution560();
        int[] nums = { 1, -1, 0 };
        int k = 0;

        int res = s.subarraySum(nums, k);

        System.out.println(res);
    }
}
